### PR TITLE
coreutils: Symlink without g prefix all non-conflicting tools

### PR DESCRIPTION
**Pull request by sjackman**
Symlink without the g prefix all tools that do not conflict with macOS.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

### CODE REVIEW

**Comment by sjackman** [comment]
This PR is one of those things that's been on the to-do list in my head for a number of years.

**Review by fxcoudert — APPROVED** [review]

File: Formula/coreutils.rb [hunk 1]
```diff
@@ -62,12 +62,17 @@ def install
...
     end
+    libexec.install_symlink "gnuman" => "man"
 
     # Symlink non-conflicting binaries
-    bin.install_symlink "grealpath" => "realpath"
-    man1.install_symlink "grealpath.1" => "realpath.1"
-
-    libexec.install_symlink "gnuman" => "man"
+    no_conflict = %w[
+      b2sum base32 chcon dir dircolors hostid md5sum nproc numfmt pinky ptx realpath runcon
```
Comment (claui) [inline Formula/coreutils.rb]:
```suggestion
      base32 chcon dir dircolors hostid md5sum nproc numfmt pinky ptx realpath runcon
```

File: Formula/coreutils.rb [hunk 2]
```diff
@@ -62,12 +62,17 @@ def install
...
+      sha1sum sha224sum sha256sum sha384sum sha512sum shred shuf stdbuf tac timeout truncate vdir
```
Comment (claui) [inline Formula/coreutils.rb]:
```suggestion
      sha224sum sha256sum sha384sum sha512sum shred shuf stdbuf tac timeout vdir
```

**Review by claui — CHANGES_REQUESTED** [review]

File: Formula/coreutils.rb [hunk 3]
```diff
@@ -62,12 +62,17 @@ def install
...
+    ]
+    no_conflict.each do |tool|
```
[the remark this answers is not in the document]
Comment (sjackman) [reply Formula/coreutils.rb]: I've changed `tool` to `cmd` for consistency.

**Review by claui — APPROVED** [review]
👍

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by sjackman** [comment]
Merged. Thanks for the reviews, all!

**Comment by ericbn** [comment]
I don't think this was a good idea. See https://github.com/Homebrew/brew/issues/5641

The idea of "non-conflicting GNU tools" might be not as simple as "any tool that BSD does not have", as for example having GNU `dircolors` (that sets `LS_COLORS`) together with BSD `ls` (which uses the `LSCOLORS` environment variable) makes no sense. I wonder which other combinations of GNU + BSD tools could be conflicting too.

**Comment by MikeMcQuaid** [comment]
> makes no sense

Makes no sense or actively breaks things?

**Comment by ericbn** [comment]
Why would we want to have GNU `dircolors` and BSD `ls` at the same time?

Let me rephrase "makes no sense": It can potentially break scripts that are made to work in both GNU and BSD by doing checks. In a script that checks for presence of (GNU) `dircolors` and figures out it exists, and then also figures out that the current `ls` is a BSD one, what should it do? Handle a third scenario where it's running in both GNU and BSD at the same time, and then what?

This enforces scripts out there to try to be much smarter than they should, having to handle all possible combinations of GNU and BSD tools in the same machine.

**Comment by sjackman** [comment]
The counter-argument is that it's simpler now to write a script that uses `sha256sum` without needing to use
```sh
sha256sum $foo || gsha256sum $foo
```
which is what I was doing previously.

**Comment by ericbn** [comment]
@sjackman, you simplified your script, and broke mine.  :- )

**Comment by ericbn** [comment]
I can make my script smarter, but I still don't see the benefit of having this mixture, instead of just offering users the ability of having all or nothing. Also you didn't answer why aren't you just adding `/usr/local/opt/coreutils/libexec/gnubin` to your path if you want to use GNU tools, as this is how this was supposed to work before your changes.

**Comment by MarcelloNicoletti** [comment]
For reference Prezto's built in Utilities module checks for `dircolors` before adding different flags to it's `ls` aliases. You can see this on [this line][prezto-util].

[prezto-util]: https://github.com/sorin-ionescu/prezto/blob/master/modules/utility/init.zsh#L74

**Comment by sjackman** [comment]
If it's specifically `dircolors`, `dir`, `vdir` that are causing grief, those symlinks could be removed.

**Comment by ericbn** [comment]
Yeah, Prezto also assumes it can do `alias ls="${aliases[ls]:-ls} --group-directories-first"` if `is-callable 'dircolors'`, so this change broke Prezto too.

**Comment by MarcelloNicoletti** [comment]
Until today, I didn't even remember I had `coreutils` installed. I got it as a requirement from something else. When I saw it as a requirement I knew I'd be fine because all of the commands it added were prefixed. I don't see why any of these should be added without prefix because it can be installed for other reasons than other than wanting GNU tools. Users like me might not expect _any_ of the GNU utilities to be installed without prefix. I'd expect all or none.

**Comment by sjackman** [comment]
The specific issue that this PR addressed is being able to easily install and run the common utilities `md5sum`, `sha1sum`, `sha256sum`, (and other `*sum`), `shuf`, `tac`, and `truncate` without additional configuration (namely adding `libexec/gnubin` to your `PATH`).

I've opened PR https://github.com/Homebrew/homebrew-core/pull/36494 to remove the symlinks for `dircolors`, `dir`, and `vdir`, due to the presence of `dircolors` breaking Prezto.
